File: clients/dart/lib/src/proof_accounts.dart
import 'package:solana_kit_addresses/solana_kit_addresses.dart';
import 'package:solana_kit_instructions/solana_kit_instructions.dart';

/// Largest proof tail accepted by the on-chain Bubblegum adapter.
const int maxBubblegumProofAccounts = 16;

/// Replaces the scalar placeholder emitted for a Pina `remaining` account with
/// the real Bubblegum proof. Empty proofs are valid for full-canopy trees.
Instruction withBubblegumProofAccounts(
  Instruction instruction,
  Iterable<Address> proofAccounts,
) {
  final proof = List<Address>.unmodifiable(proofAccounts);
  if (proof.length > maxBubblegumProofAccounts) {
    throw RangeError.range(
      proof.length,
      0,
      maxBubblegumProofAccounts,
      'proofAccounts',
    );
  }
  final accounts = instruction.accounts;
  if (accounts == null || accounts.isEmpty) {
    throw StateError('instruction has no generated proof placeholder');
  }
  return Instruction(
    programAddress: instruction.programAddress,
    accounts: <AccountMeta>[
      ...accounts.take(accounts.length - 1),
      for (final address in proof)
        AccountMeta(address: address, role: AccountRole.readonly),
    ],
    data: instruction.data,
  );
}

File: clients/typescript/lootbox_program/src/proofAccounts.ts
import {
	AccountRole,
	type AccountMeta,
	type Address,
	type Instruction,
	type InstructionWithAccounts,
} from "@solana/kit";

type InstructionWithAccountList<TProgram extends string> = Instruction<TProgram> &
	InstructionWithAccounts<readonly AccountMeta[]>;

/** Largest proof tail accepted by the on-chain Bubblegum adapter. */
export const MAX_BUBBLEGUM_PROOF_ACCOUNTS = 16;

/**
 * Replace the scalar placeholder emitted for a Pina `remaining` account with
 * the real Bubblegum proof. Empty proofs are valid for full-canopy trees.
 */
export function withBubblegumProofAccounts<TProgram extends string>(
	instruction: InstructionWithAccountList<TProgram>,
	proofAccounts: readonly Address[],
): InstructionWithAccountList<TProgram> {
	if (proofAccounts.length > MAX_BUBBLEGUM_PROOF_ACCOUNTS) {
		throw new RangeError(
			`Bubblegum proofs may contain at most ${MAX_BUBBLEGUM_PROOF_ACCOUNTS} accounts`,
		);
	}
	if (instruction.accounts.length === 0) {
		throw new RangeError("instruction has no generated proof placeholder");
	}
	return Object.freeze({
		...instruction,
		accounts: Object.freeze([
			...instruction.accounts.slice(0, -1),
			...proofAccounts.map((address) =>
				Object.freeze({ address, role: AccountRole.READONLY })
			),
		]),
	});
}
